### ISSUE TITLE
Separate compilation violates package-protected accessibility

### ISSUE THREAD

**Issue by eed3si9n** [scala/bug#12321]
This is a retroactive bug report on Scala 2.12.x and 2.13.0 ~ 1 since it seems to be fixed in Scala 2.13.2.

## reproduction steps

using Scala 2.13.1

### bar/B.scala

```scala
package bar

protected[bar] class B(x: Int)
```

### foo/Test.scala

```scala
package example

import bar.B

object Test {
  def b = new B(1)
}
```

### bash

```bash
mkdir -p bar/classes/
mkdir -p foo/classes/
$HOME/Applications/scala-2.13.1/bin/scalac bar/B.scala -d bar/classes
ls bar/classes/bar
# outputs B.class

$HOME/Applications/scala-2.13.1/bin/scalac -cp bar/classes/ foo/Test.scala -d foo/classes
ls foo/classes/example
# outputs Test$.class  Test.class
```

## problem

The separate compilation violates `protected[bar]`.

## expectation

`Test` should not compile.

```scala
foo/Test.scala:6: error: not found: type B
  def b = new B(1)
              ^
1 error
```

## note

If I compile all sources together the accessibility is checked correctly:

```
$ $HOME/Applications/scala-2.13.1/bin/scalac bar/B.scala  foo/Test.scala -d foo/classes
foo/Test.scala:6: error: not found: type B
  def b = new B(1)
              ^
one error found
```

## keyword for search

visibility

**Comment by som-snytt** [comment]
Duplicates https://github.com/scala/bug/issues/8121 which also turns on an import and separate compilation.

Fix was https://github.com/scala/scala/pull/8753

I thought there was a similar "qualified access" ticket, but maybe not. It deserves a regression test, since Lukas feared side effects of symbol forcing, but I don't think he necessarily anticipated that other checks would start to work correctly.

**Comment by eed3si9n** [comment]
@som-snytt Thanks for the pointers!

Here's a PR for tests - https://github.com/scala/scala/pull/9453
